### PR TITLE
fix inaccessible code blocks

### LINKED ISSUE #10335: Code Snippets are not accessible
**Describe the bug**
On desktop, user cannot navigate to the code snippet using keyboard only. User cannot put the snippet in focus, cannot scroll through it or have it read by 

**To Reproduce**

1. Create a new post.
2. Insert a code snippet with a long line - for example : 
```JavaScript
console.log("Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. It has survived not only five centuries, but also the leap into electronic typesetting, remaining essentially unchanged. It was popularised in the 1960s with the release of Letraset sheets containing Lorem Ipsum passages, and more recently with desktop publishing software like Aldus PageMaker including versions of Lorem Ipsum
```
3. Publish.
4. On desktop try do navigate to the snippet using keyboard. You can't.
5. Attempt to scroll through to the end of the snippet - you can't.
6. This also means that the code snippets are not read by a screen reader.

**Expected behavior**

1. User should be able to navigate to a code snippet using a keyboard
2. User should be able to interact with the code snippet using the keyboard
3. Screen Reader should pick up 

**Screenshots**

Not needed.

**Desktop (please complete the following information):**

- OS, version:
- Browser, version:

**Smartphone (please complete the following information):**

- Device: Dell Laptop
- OS, version: Windows
- Browser, version: Version 85.0.4183.102 (Official Build) (64-bit)

**Additional context**

### PR DESCRIPTION
**Pull request by Amorpheuz**
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [ ] Feature
- [x] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description

Adds the following changes:
- add `tabindex = "0"` to `pre` tags.
- add border color change on focus to highlight focus.
- update tests.

I added `tabindex` to `pre` tags since I assumed that all kinds of code snippets should exhibit similar behavior. If there are some other features breaking due to this, I can switch to a more specific version.

## Related Tickets & Documents

Fixes #10335 

## QA Instructions, Screenshots, Recordings

- How a code snippet is highlighted:
image

I tested it with following types of code snippets: enclosed in triple `\`` with and without language specified, code snippet due to indentation, code snippet in lists. I would also like feedback on if I updated the test correctly or not.

## Added tests?

- [x] yes
- [ ] no, because they aren't needed
- [ ] no, because I need help

## Added to documentation?

- [ ] docs.forem.com
- [ ] readme
- [x] no documentation needed

## [optional] Are there any post deployment tasks we need to perform?

Should require re-rendering of posts to make the fix backward compatible.

## [optional] What gif best describes this PR or how it makes you feel?

Like every year, October being the exam month made me delay the PR 😕, but finally I am free until the next one.

Happy Face on a pumpkin

### CODE REVIEW

**Comment by Amorpheuz** [comment]
Hey, I would also like to add accessibility with the new-full screen functionality by making it keyboard navigable. Is it something useful? Should I update this PR with the same?

**Comment by nickytonline** [comment]
I'm not sure about the accessibility story for this. `<pre />` tags don't appear to be focusable out of the box. It's true that adding a `tabIndex` will make it focusable, but I'm not sure if this is what we should do. 

Based on https://www.w3.org/html/wiki/Elements/pre, they say this:

<quote>"The author should consider accessibility, when use the pre element. This is because, when speech synthesizers, braille displays, and the like is used, there is a possibility that preformatted text is destroyed. For exapmle, For cases like ASCII art, it is likely that an alternative presentation, such as a textual description, would be more universally accessible to the readers of the document."<quote>

Maybe @colabottles has some input here.

**Comment by reobin** [comment]
I would argue that what should be focusable is actually the full-screen button (which is not focusable right now).

The `:focus-within` style and the overall functionality @Amorpheuz worked on would still work as intended, and it would have the added benefit of making the full-screen button keyboard accessible.

Maybe we could look into turning the full-screen button into an actual `<button>`?

**Comment by nickytonline** [comment]
> I would argue that what should be focusable is actually the full-screen button (which is not focusable right now).
> 
> The `:focus-within` style and the overall functionality @Amorpheuz worked on would still work as intended, and it would have the added benefit of making the full-screen button keyboard accessible.
> 
> Maybe we could look into turning the full-screen button into an actual `<button>`?

I don't think I saw that PR when it got merged. If it's not a native button, making it one is definitely a good call @reobin.

**Comment by colabottles** [comment]
> I'm not sure about the accessibility story for this. `<pre />` tags don't appear to be focusable out of the box. It's true that adding a `tabIndex` will make it focusable, but I'm not sure if this is what we should do.
> 
> Based on https://www.w3.org/html/wiki/Elements/pre, they say this:
> 
> "The author should consider accessibility, when use the pre element. This is because, when speech synthesizers, braille displays, and the like is used, there is a possibility that preformatted text is destroyed. For exapmle, For cases like ASCII art, it is likely that an alternative presentation, such as a textual description, would be more universally accessible to the readers of the document."
> 
> Maybe @colabottles has some input here.

You can wrap the `<pre>` in a `figure` tag and it should be accessible via keyboard & screen reader. You can even wrap it into a `section` tag as well.

```
<figure>
  <figcaption id="example1-caption">
    This is the caption
  </figcaption>
  <pre aria-labelledby="example1-caption" aria-describedby="example1-description">
    <code>function getToSleep()
      {
        while(noise &lt;= 10 &amp;&amp; sleep !== "zzz")
      {
        sheep ++;
      }
        if(noise &gt; 10)
      {
        return false;
      }
        return true;
      }</code>
   </pre>
</figure>
```

`tabIndex` is also a viable approach as well.

**Comment by colabottles** [comment]
If it were my choice, wrapping the `<pre><code>...</code></pre>` block in a `section` tag with aria labels.

**Comment by colabottles** [comment]
> I would argue that what should be focusable is actually the full-screen button (which is not focusable right now).
> 
> The `:focus-within` style and the overall functionality @Amorpheuz worked on would still work as intended, and it would have the added benefit of making the full-screen button keyboard accessible.
> 
> Maybe we could look into turning the full-screen button into an actual `<button>`?

Agree with this 100%, which brings to mind this from a tweet today by Andy Bell.

> - If it goes somewhere: use an `<a>`
> - If it does something: use a `<button>`
> - If it does either: don’t use a `<div>`

**Comment by nickytonline** [comment]
Thanks for your feedback @colabottles! 👏🏻

**Comment by reobin** [comment]
Quick demo of what it would look like:

demo

The fullscreen button is turned into a button, which makes the code block "focusable", and also allows the user to use `Enter` to go in and out of fullscreen when the code block is "focused".

If that's the way to go, I can help out @Amorpheuz to achieve this. Not many changes were necessary from what they had already.

**Comment by Amorpheuz** [comment]
Thanks for the help @colabottles @nickytonline @reobin , I was planning to switch the `div` to a `button` as well to make it more accessible. I didn't know `tabIndex` could have such side-effects on screen readers though. I will have to test out @reobin's PR for the same and mark the PR ready for review once done.

P.S. Thanks for the addition Reobin <3

**Comment by Amorpheuz** [comment]
The code block highlight works like a charm, but the original purpose of the issue #10335 doesn't get fixed (no scroll with keyboard keys) by the change since due to how the code-block is structured. I think I will be implementing the structure that takes benefit of `section` as suggested.

Another solution that came to mind was reading keypress events on the button and scrolling the window via js but I feel thats adding unecessary complexity.

**Comment by reobin** [comment]
@Amorpheuz Like you said, I think the solution @colabottles mentioned would work best here. If you need help, don't hesitate to reach out :)

Something I don't quite get from the original issue is the scroll problem. How exactly should a code block be scrollable with the keyboard? Is there an accessibility shortcut I'm missing that lets you scroll down a focused element until you reach its bottom?

Regardless, the code block's content should definitely be screenreader-friendly so let's work on that

**Comment by nickytonline** [comment]
Putting this to draft as it is still a wip.

**Comment by Amorpheuz** [comment]
@reobin Since the container can overflow when a lengthy line of code is present in a single line, it makes the code block scrollable horizontally. Now to make it scroll horizontally with <kbd>←</kbd> and <kbd>→</kbd> keyboard arrows, it requires the element to be focusable. Once the element is focused, the code block will become accessible by keyboard.

I found more about this from an article on [Keyboard Accessibility by WebAIM](https://webaim.org/techniques/keyboard/#testing).

I will post here if I hit a roadblock!

**Comment by colabottles** [comment]
> @Amorpheuz Like you said, I think the solution @colabottles mentioned would work best here. If you need help, don't hesitate to reach out :)
> 
> Something I don't quite get from the original issue is the scroll problem. How exactly should a code block be scrollable with the keyboard? Is there an accessibility shortcut I'm missing that lets you scroll down a focused element until you reach its bottom?
> 
> I'm referring to this line from #10335:
> 
> > 1. Attempt to scroll through to the end of the snippet - you can't.
> 
> Regardless, the code block's content should definitely be screenreader-friendly so let's work on that

I will refer you to this quick and dirty experiment that I threw together here https://cdpn.io/colabottles/debug/dyXdrag to test. Mind you, VoiceOver in Firefox is still being worked on, so best bet is to test in Safari since VO was made to work with Safari. Chrome still has some bugs. You can access the pen here: https://codepen.io/colabottles/pen/dyXdrag

**Comment by Amorpheuz** [comment]
Hey @nickytonline I wanted some feedback on the structure I planned out after the suggestions before I push it. Here's what I figured:

```html
<div class="highlight"> <!-- wrapper for styling, done by rouge -->
  <section aria-labelledby="codeblock-0" tabindx="0"> <!-- what I add to increase  screenreader and keyboard accessibility -->
     <label id="codeblock-0">
       <!-- Language (e.g. YAML) --> Code Block
     </label>
     <pre class="highlight"> <!-- wrapper for code itself, done by rouge -->
        <code>
           <!-- some fancy code -->
        </code>
     </pre>
  </section>
  <div class="highlight__panel js-actions-panel">
    <button class="highlight__panel-action js-fullscreen-code-action">
      <svg> <!-- fullscreen icon svg --> </svg>
    </button>
  </div>
</div>
```

A question I had was which `role` attribute value should be used to describe the section, since a **landmark/widget role** is required in order for the `aria` attributes to work?

I can ensure that CSS changes are done in a way that the past code-blocks don't break themselves in case re-rendering articles is not possible, which I guess should be the case considering the massive number of articles present.

> P.S. Sorry about the messed up file changes, I will mistakenly pushed my rebase that I did for testing. I will rollback the rebase / open a new PR once changes have been finalised in order to allow you to better review my code.

**Comment by citizen428** [comment]
@nickytonline Drawing your attention back to this PR, from the last comment it seems like the ball is in our court.

**Comment by nickytonline** [comment]
> Hey @nickytonline I wanted some feedback on the structure I planned out after the suggestions before I push it. Here's what I figured:
> 
> ```
> <div class="highlight"> 
>   <section aria-labelledby="codeblock-0" tabindx="0"> 
>      <label id="codeblock-0">
>         Code Block
>      </label>
>      <pre class="highlight"> 
>         <code>
>            
>         </code>
>      </pre>
>   </section>
>   <div class="highlight__panel js-actions-panel">
>     <button class="highlight__panel-action js-fullscreen-code-action"> 
>       <svg>  </svg>
>     </button>
>   </div>
> </div>
> ```
> 
> A question I had was which `role` attribute value should be used to describe the section, since a **landmark/widget role** is required in order for the `aria` attributes to work?
> 

I think this structure makes sense based on @colabottles feedback. For the full-screen icon, make sure it has a title or the button has some hidden text to indicate what it does.

I'm also going to defer to @aitchiss to see if she has any concerns in regards to accessibility.

**Comment by aitchiss** [comment]
I think we can simplify your proposed structure slightly @Amorpheuz:

```html
<div class="highlight"> 
     <span id="codeblock-0" class="screen-reader-only">
       <!-- Language (e.g. YAML) --> Code Block
     </span>
     <pre class="highlight" tabindex="0" aria-labelledby="codeblock-0"> <!-- wrapper for code itself, done by rouge -->
        <code>
           <!-- some fancy code -->
        </code>
     </pre>
  <div class="highlight__panel js-actions-panel">
    <button class="highlight__panel-action js-fullscreen-code-action">
      <svg> <!-- fullscreen icon svg --> </svg>
    </button>
  </div>
</div>
```

All I've changed is:
- Removed the section. Sections are generally standalone sections of a page, and should normally have a heading inside them. I think in the case of code blocks, they usually illustrate a point made elsewhere in a post, and so don't "stand alone" in the document.
- Changed the `label` to a `span` since labels are normally used to label form inputs, which we don't have here. I've also noted the class `screen-reader-only` on the span - this is a utility class we have that makes sure it's hidden from view in the UI
- Moved the `tabindex` and `aria-labelledby` to the `pre` element

Let me know if you have any questions about this! 


**Some other issues:**

There are some further problems with the "full screen" feature, in that we don't trap the focus inside the overlayed code sample - you can continue to tab through items behind the full screen code, and even if we make the toggle button a real button, it will be difficult to reach by keyboard without a focus trap.

The scope of this fix is already creeping though, so I'm happy if we want to address that particular issue in a separate PR.

**Comment by nickytonline** [comment]
Good point on the label @aitchiss and thanks for clarifying about section.

For the focus trap, I would create a separate issue and we can dynamically import it like we do for modals. There might be a bit more to it as this is currently not a Preact component.

**Comment by Amorpheuz** [comment]
Hey @aitchiss, I have been having setup issues with the installation where the database setup is erroring out with `Missing host to link to!......` error while running `bin/setup`. I have been unable to investigate more into this due to some things going on irl.

I will post a more detailed error log later today and hopefully that should help sort out the issue. Thanks for reaching out!

**Comment by aitchiss** [comment]
> Hey @aitchiss, I have been having setup issues with the installation where the database setup is erroring out with `Missing host to link to!......` error while running `bin/setup`. I have been unable to investigate more into this due to some things going on irl.
> 
> I will post a more detailed error log later today and hopefully that should help sort out the issue. Thanks for reaching out!

No problem! Let us know more detail on the error and I'm sure we'll be able to help 🙂

**Comment by Amorpheuz** [comment]
Hey @aitchiss. here's the error I get while running `bin/setup` during the preparing database step:

```
== Preparing database ==
Created database 'Forem_development'
Created database 'Forem_test'
WARNING: sqlite and mysql triggers may not be shared by multiple actions
Seeding with multiplication factor: 1

  1. Creating Organizations.
  2. Creating DEV profile fields.
  3. Creating 10 Users.
2021-06-06T08:01:47.408Z pid=876 tid=1nxo uniquejobs=client until_executing=uniquejobs:b897a2a141f458357b09fe48ac4f4909 WARN: Timed out after 0s while waiting for primed token (digest: uniquejobs:b897a2a141f458357b09fe48ac4f4909, job_id: 1f26e3600b1573a23befd9c2)
rails aborted!
ActionView::Template::Error: Missing host to link to! Please provide the :host parameter, set default_url_options[:host], or set :only_path to true
/home/yash/projects/forem/app/views/devise/mailer/confirmation_instructions.html.erb:5:in `_app_views_devise_mailer_confirmation_instructions_html_erb__1517433354806972399_108540'
/home/yash/projects/forem/db/seeds.rb:69:in `block (2 levels) in <main>'
/home/yash/projects/forem/db/seeds.rb:66:in `times'
/home/yash/projects/forem/db/seeds.rb:66:in `block in <main>'
/home/yash/projects/forem/app/lib/seeder.rb:30:in `create_if_none'
/home/yash/projects/forem/db/seeds.rb:63:in `<main>'
/home/yash/projects/forem/bin/rails:5:in `<top (required)>'
/home/yash/projects/forem/bin/spring:10:in `require'
/home/yash/projects/forem/bin/spring:10:in `block in <top (required)>'
/home/yash/projects/forem/bin/spring:7:in `tap'
/home/yash/projects/forem/bin/spring:7:in `<top (required)>'

Caused by:
ArgumentError: Missing host to link to! Please provide the :host parameter, set default_url_options[:host], or set :only_path to true
/home/yash/projects/forem/app/views/devise/mailer/confirmation_instructions.html.erb:5:in `_app_views_devise_mailer_confirmation_instructions_html_erb__1517433354806972399_108540'
/home/yash/projects/forem/db/seeds.rb:69:in `block (2 levels) in <main>'
/home/yash/projects/forem/db/seeds.rb:66:in `times'
/home/yash/projects/forem/db/seeds.rb:66:in `block in <main>'
/home/yash/projects/forem/app/lib/seeder.rb:30:in `create_if_none'
/home/yash/projects/forem/db/seeds.rb:63:in `<main>'
/home/yash/projects/forem/bin/rails:5:in `<top (required)>'
/home/yash/projects/forem/bin/spring:10:in `require'
/home/yash/projects/forem/bin/spring:10:in `block in <top (required)>'
/home/yash/projects/forem/bin/spring:7:in `tap'
/home/yash/projects/forem/bin/spring:7:in `<top (required)>'
Tasks: TOP => db:prepare
(See full trace by running task with --trace)

== Command ["bin/rails app_initializer:setup forem:setup"] failed ==
```

I am running this on WSL 2 + Ubuntu 18.04 LTS, PostgreSQL 13.3, Ruby 2.7.2, Node 14.17.0. Let me know if I should share any more details that I might have missed!

**Comment by aitchiss** [comment]
Hey @Amorpheuz - do you have your `.env` file set up? The "host" referred to in that error message is taken from an environment variable in that file.

In most cases all you would need is to copy the `.env_sample` file and name it `.env`. This should contain all the environment variables you need, including the `APP_DOMAIN` which transaltes to the "host" noted in the error above.

Let me know if that helps at all!

**Comment by Amorpheuz** [comment]
Hey @aitchiss, thanks for the help. It seems that was indeed the error. I or something in my editor had messed up the `.env` file which I failed to notice. I re-created the same file again from the sample file and now it seems to be working fully!

I hope to get back to you with the updated pr during this weekend, thanks! 😃

**Comment by aitchiss** [comment]
@Amorpheuz awesome! I'm so pleased to hear it's all working now 😄

**Comment by citizen428** [comment]
Hi @Amorpheuz, just checking in to see if you need any other help from us? 😃

**Comment by Amorpheuz** [comment]
@citizen428 Hey, thanks for reaching out. It took me some time to refamiliarize myself with the codebase since things related to the PR had moved locations since opening the PR. During the past weekend, I was finally able to retrace my steps up to the latest point in the PR and I have pushed the same to a new temporary branch on my fork. You can find the changes here: https://github.com/Amorpheuz/forem/pull/5

I had a couple of doubts that I would like help with:

1. What would be a safe way to update the current branch with the more updated version? Would a simple rebase suffice?
2. In reference to the structure defined by `@aitchiss`  in https://github.com/forem/forem/pull/11058#issuecomment-837990530, what would be an effective way to extract the language of the code-block from? And what would be a safe way to guarantee unique ids for the `<pre>` blocks?

I am having trouble recalling what I was planning to use in my original approach, since it has been some time, but I don't think a simple `codeblock-0` numbering would suffice since the parser would be run through for each user generated content on the page and that might include code blocks in the comments for which the numbering should restart from 0 (for every comment) if I am understanding this correctly. 

Thanks for the help!

**Comment by citizen428** [comment]
Re 1: you can either update this PR or close it and open a new one, whatever is easier for you.
Re 2: I don't know off the top of my hat and would need to familiarize myself with the current status first, which unfortunately I don't have time for right now. 😢 Maybe someone else can jump in and help on this, otherwise, I'll try to get to it early next week.

File: app/assets/stylesheets/components/syntax.scss [hunk 1]
```diff
@@ -68,6 +81,9 @@ div.highlight {
   &:hover {
     opacity: 1;
   }
+  &:focus {
+    outline: none;
```
Comment (nickytonline) [inline app/assets/stylesheets/components/syntax.scss:85]: @aitchiss, I imagine we'd still need some kind of visual indicator that this has focus or stay with the default via focus-within.

**Comment by nickytonline** [comment]
> @citizen428 Hey, thanks for reaching out. It took me some time to refamiliarize myself with the codebase since things related to the PR had moved locations since opening the PR. During the past weekend, I was finally able to retrace my steps up to the latest point in the PR and I have pushed the same to a new temporary branch on my fork. You can find the changes here: [Amorpheuz#5](https://github.com/Amorpheuz/forem/pull/5)
> 
> I had a couple of doubts that I would like help with:
> 
> 1. What would be a safe way to update the current branch with the more updated version? Would a simple rebase suffice?
> 2. In reference to the structure defined by `@aitchiss`  in [fix inaccessible code blocks #11058 (comment)](https://github.com/forem/forem/pull/11058#issuecomment-837990530), what would be an effective way to extract the language of the code-block from? And what would be a safe way to guarantee unique ids for the `<pre>` blocks?
> 
> I am having trouble recalling what I was planning to use in my original approach, since it has been some time, but I don't think a simple `codeblock-0` numbering would suffice since the parser would be run through for each user generated content on the page and that might include code blocks in the comments for which the numbering should restart from 0 (for every comment) if I am understanding this correctly.
> 
> Thanks for the help!

For the counter for the code blocks, I'm not sure yet, but for the rest, the HTML structure that @aitchiss proposed, makes sense. In terms of updating the branch, doing a merge should be fine to get you up to date.

**Comment by citizen428** [comment]
Hi @Amorpheuz 👋 Are you still working on this PR or should we close it?

**Comment by Amorpheuz** [comment]
Hey @citizen428, I have been a bit busy since past few weeks as I completed my university and found a new job. I don't think, I will be able to focus on this PR for sometime into the future as well. Sorry for that. 😅

Do close this PR and re-assign the issue as required! I would love to see this feature to Forem.

**Comment by citizen428** [comment]
That's ok, @Amorpheuz. Thanks for your contribution and good luck with the new job! 😃

**Pull request closed without merging**